query.go

```diff
@@ -124,12 +119,23 @@ func newQueryRunner(q *dhtQuery) *dhtQueryRunner {
 	return r
 }
 
-func (r *dhtQueryRunner) Run(ctx context.Context, peers []peer.ID) (*dhtQueryResult, error) {
+func (r *dhtQueryRunner) run(ctx context.Context, peers []peer.ID) ([]peer.ID, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	r.log = logger
 	r.runCtx = ctx
 
+	err := r.recurse(ctx, peers)
+	if err != nil {
+		return nil, err
+	}
+	return r.finish(ctx)
+}
+
+func (r *dhtQueryRunner) recurse(ctx context.Context, peers []peer.ID) error {
 	// setup concurrency rate limiting
-	for i := 0; i < r.query.concurrency; i++ {
+	for len(r.rateLimit) < cap(r.rateLimit) {
 		r.rateLimit <- struct{}{}
```

Comment:
what is this doing exactly? Are we just trying to fill up the channel?

Reply:
Yes (unnecessary but I was already messing with this code).